mixer: enforce type discipline for attribute values

kyessenov opened the pull request:
- enforce `StringMap` use as the attribute value across the code base
- add `List`, used only by the proto decoder (otherwise, we would have to disable that code) and not surfaced to the user
- set L2 cache size to 0

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/10260#" title="Author self-approved">kyessenov</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **mandarjog**

If they are not already assigned, you can assign the PR to them by writing `/assign @mandarjog` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[install/OWNERS](https://github.com/istio/istio/blob/release-1.1/install/OWNERS)**
- ~~[mixer/OWNERS](https://github.com/istio/istio/blob/release-1.1/mixer/OWNERS)~~ [kyessenov]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mandarjog"]} -->

geeknoid commented on mixer/pkg/attribute/bag.go:
```diff
@@ -32,3 +52,42 @@ type Bag interface {
+		}
+	case StringMap:
+		if y, ok := that.(StringMap); ok {
+			return x.Equal(y)
```
Shouldn't this use relfect.DeepEquals like compareAttributes was?

kyessenov replied:
It is DeepEquals under the hood. The intent is to compare as-value, ignoring the parent proto bag.

kyessenov replied:
The new function is equivalent to the old one and made public (for use beyond test utils)

kyessenov commented:
This fixes the specific bug pointed out in #9596 . However, we are still disabling the cache, until we install  regression testing to prevent this error from happening in the future.

geeknoid commented on mixer/pkg/attribute/list.go:
```diff
@@ -0,0 +1,62 @@
+	}
+}
+
+// Append adds an element to the list
```
This comment is a bit misleading. It's not appending to the list, it's producing a new list which is a concatenation of the old list with a new element.

Also, although this seem like a functional immutable copy, this is not so given the semantics of the underlying slice. If you have l1:

  l2 = l1.Append(x)
  l3 = l1.Append(x)

This will not work as you'd expect with a purely functional design. 

So it might be cleaner to just have Append take a (l*List) and mutate it in place without returning a List.

kyessenov replied:
Good point. I changed it to take a pointer and mutate in place.

istio-testing commented:
@kyessenov: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/191) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

geeknoid commented on mixer/pkg/attribute/list.go:
```diff
@@ -0,0 +1,62 @@
+	}
+}
+
+// String prints the entries in the list
```
Doesn't actually print, just returns a string.

kyessenov replied:
Rephrased.

geeknoid commented on mixer/pkg/attribute/list.go:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Istio Authors
```
So what is this type used for? Is this to implement eventual support for a STRING_LIST attribute type?

kyessenov replied:
Yes, the protobuf decoder relies on ability to put lists of repeated values. This is a hatch to keep the decoder aware of repeated vales, without fully implementing the list support (it's beyond the scope).

geeknoid commented on mixer/pkg/attribute/stringmap.go:
```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Istio Authors
```
I think you need simple unit tests for the list and stringmap types.

kyessenov replied:
Yes! I'll add them. Want to make sure coverage is actually working in CI.

geeknoid commented on mixer/pkg/attribute/stringmap.go:
```diff
@@ -0,0 +1,88 @@
+func (s StringMap) Get(key string) (string, bool) {
+	str, found := s.entries[key]
+
+	// the string map may be de-touched from the owning bag
```
de-touched -> detached

kyessenov replied:
Thanks. Done.

geeknoid commented on mixer/pkg/attribute/stringmap.go:
```diff
@@ -0,0 +1,88 @@
+	return StringMap{name: s.name, entries: c, pb: s.pb}
+}
+
+// String prints the entries in the string map
```
Doesn't print

kyessenov replied:
Done.

geeknoid commented on mixer/pkg/attribute/mutableBag.go:
```diff
@@ -69,9 +69,16 @@ func GetMutableBag(parent Bag) *MutableBag {
-	m.values = v
+	m.values = values
+	for k, v := range values {
+		switch v.(type) {
```
Can you use the CheckType function here?

geeknoid commented on mixer/pkg/server/args.go:
```diff
@@ -124,7 +124,7 @@ func DefaultArgs() *Args {
 		IntrospectionOptions:   ctrlz.DefaultOptions(),
 		EnableProfiling:        true,
-		NumCheckCacheEntries:   5000 * 5 * 60, // 5000 QPS with average TTL of 5 minutes
+		NumCheckCacheEntries:   0, // prior value 5000 * 5 * 60, 5000 QPS with average TTL of 5 minutes
```
I think you should this as is.

Instead, set the value to 0 explicitly right before creating the cache. We want it to be disabled no matter what the user says on the command-line.

geeknoid commented on install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml:
```diff
@@ -28,6 +28,7 @@
           - --monitoringPort={{ .Values.global.monitoringPort }}
           - --address
           - unix:///sock/mixer.socket
+          - --numCheckCacheEntries=0
```
I don't think we want to add this here. We want to forcibly disable the cache, no matter what the user specifies. So I think we should just in the code and right at the point where the cache is being configure, we explicit pass 0 there and ignore what the user specifies on the command-line.

kyessenov commented:
Revewers: I'm not fully done, but I want to use the CI machinery to validate e2e behavior with my change. I suspect there's some dependency on the check cache somewhere, but I'm not sure where...

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/191) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth-mcp.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth-mcp/2860) | `/test e2e-mixer-no_auth-mcp`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/191) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth-mcp.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth-mcp/2860) | `/test e2e-mixer-no_auth-mcp`
prow/e2e-mixer-no_auth.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8948) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/10260#" title="Author self-approved">kyessenov</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approvers: **mandarjog**, **nmittler**

If they are not already assigned, you can assign the PR to them by writing `/assign @mandarjog @nmittler` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[install/OWNERS](https://github.com/istio/istio/blob/release-1.1/install/OWNERS)**
- ~~[mixer/OWNERS](https://github.com/istio/istio/blob/release-1.1/mixer/OWNERS)~~ [kyessenov]
- **[tests/OWNERS](https://github.com/istio/istio/blob/release-1.1/tests/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mandarjog","nmittler"]} -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth-mcp.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth-mcp/2860) | `/test e2e-mixer-no_auth-mcp`
prow/e2e-mixer-no_auth.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8948) | `/test e2e-mixer-no_auth`
prow//e2e-simpleTests-minProfile.sh | db4eb2c6bb4d1ab5c77ad85fd8665415d6e138f9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/194) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth-mcp.sh | b9f1c15d767d523ab3c085f025e4d89c7bc97592 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth-mcp/2860) | `/test e2e-mixer-no_auth-mcp`
prow//e2e-simpleTests-minProfile.sh | db4eb2c6bb4d1ab5c77ad85fd8665415d6e138f9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/194) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth.sh | db4eb2c6bb4d1ab5c77ad85fd8665415d6e138f9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8953) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/10260#" title="Author self-approved">kyessenov</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **hklai**

If they are not already assigned, you can assign the PR to them by writing `/assign @hklai` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- ~~[mixer/OWNERS](https://github.com/istio/istio/blob/release-1.1/mixer/OWNERS)~~ [kyessenov]
- **[tests/OWNERS](https://github.com/istio/istio/blob/release-1.1/tests/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["hklai"]} -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth.sh | db4eb2c6bb4d1ab5c77ad85fd8665415d6e138f9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8953) | `/test e2e-mixer-no_auth`
prow//e2e-simpleTests-minProfile.sh | f820b12df000841844928f175002303ad40c493b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/197) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth.sh | db4eb2c6bb4d1ab5c77ad85fd8665415d6e138f9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8953) | `/test e2e-mixer-no_auth`
prow//e2e-simpleTests-minProfile.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/198) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/198) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8957) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-mixer-no_auth

Seems something is flaky there...

```
mixer_test.go:843: Query prometheus to get baseline cache hits...
mixer_test.go:849: Baseline cache hits: 15
mixer_test.go:850: Start to call visit function...
mixer_test.go:856: Query promethus to get new cache hits number...
mixer_test.go:862: New cache hits: 15
mixer_test.go:867: Check cache hit: 0 is less than expected: 1			
```

kyessenov commented:
PTAL, the PR is ready for review.

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/198) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8960) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-mixer-no_auth

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/198) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8961) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8961) | `/test e2e-mixer-no_auth`
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-mixer-no_auth.sh | e8b60382f53fef3016316c39433e8c6ee883a67b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8961) | `/test e2e-mixer-no_auth`
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-dashboard.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8690) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-dashboard

douglas-reid commented on mixer/pkg/attribute/bag.go:
```diff
@@ -14,9 +14,29 @@
+// - time.Time
+// - time.Duration
+// - []byte (backed by a byte array)
+// - attribute.StringMap (backed by a map[string]string)
```
:+1:

douglas-reid commented on mixer/pkg/attribute/bag.go:
```diff
@@ -32,3 +52,42 @@ type Bag interface {
+}
+
+// CheckType validates that an attribute value has a supported type.
+func CheckType(value interface{}) bool {
```
:+1:

douglas-reid commented on mixer/pkg/attribute/stringmap_test.go:
```diff
@@ -0,0 +1,33 @@
+
+	"istio.io/istio/mixer/pkg/attribute"
+)
+
```
should we test `Get()` here?

kyessenov replied:
It is implicitly tested by the big `bag_test`. These are edge cases to improve the coverage.

douglas-reid commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -1323,7 +1327,7 @@ func allowRuleSync() {
 func allowPrometheusSync() {
 	log.Info("Sleeping to allow prometheus to record metrics...")
-	time.Sleep(15 * time.Second)
+	time.Sleep(30 * time.Second)
```
we had just turned this down to 15 in an effort to speed up tests (prom scrape time was reduced to 1s). Have you seen issues around timing here?

kyessenov replied:
It failed twice, once on circle and once on prow. 30s seems to pass.

kyessenov commented:
Updated the PR to include fixes for the code generation. It used to not wrap net.IP and map[string]string in the right boxes before placing into the bags.

I also replaced `int` with `int64` everywhere in the bag.

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/e2e-dashboard.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8719) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/e2e-dashboard.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8719) | `/test e2e-dashboard`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/11959) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/e2e-dashboard.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8719) | `/test e2e-dashboard`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/11959) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8989) | `/test e2e-mixer-no_auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/e2e-dashboard.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8719) | `/test e2e-dashboard`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/11959) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8989) | `/test e2e-mixer-no_auth`
prow/istio-pilot-multicluster-e2e.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5694) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

geeknoid commented on mixer/pkg/attribute/bag_test.go:
```diff
@@ -37,8 +36,8 @@ var (
 	sm2 := mixerpb.StringMap{Entries: map[int32]int32{-17: -17}}
-	m1 := map[string]string{"N16": "N16"}
-	m3 := map[string]string{"N42": "FourtyTwo"}
+	m1 := StringMap{entries: map[string]string{"N16": "N16"}}
```
Worth using the WrapStringMap function here and in other similar places?

kyessenov replied:
Done

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTests/14075) | `/test e2e-simpleTests`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/e2e-dashboard.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8719) | `/test e2e-dashboard`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/11959) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8989) | `/test e2e-mixer-no_auth`
prow/istio-pilot-multicluster-e2e.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5694) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-local-tests.sh | 39c3eb7a06de4111d841dc18f8e5a378f7b678bd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-integ-local-tests/1647) | `/test istio-integ-local-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-bookInfoTests-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-bookInfoTests-envoyv2-v1alpha3/11166) | `/test e2e-bookInfoTests-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/11959) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/e2e-mixer-no_auth.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8989) | `/test e2e-mixer-no_auth`
prow/istio-pilot-multicluster-e2e.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5694) | `/test istio-pilot-multicluster-e2e`
prow/e2e-dashboard.sh | 0e5164b140306b1eec2ccdb09fa5f39c3c09f807 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8736) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-dashboard

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-mixer-no_auth.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-mixer-no_auth/8989) | `/test e2e-mixer-no_auth`
prow/istio-pilot-multicluster-e2e.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5694) | `/test istio-pilot-multicluster-e2e`
prow/e2e-dashboard.sh | 0e5164b140306b1eec2ccdb09fa5f39c3c09f807 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8738) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-dashboard

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/istio-pilot-multicluster-e2e.sh | 4226be8b5e960d736ce8ee757fd4084705337120 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5694) | `/test istio-pilot-multicluster-e2e`
prow/e2e-dashboard.sh | 0e5164b140306b1eec2ccdb09fa5f39c3c09f807 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8739) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-dashboard.sh | 0e5164b140306b1eec2ccdb09fa5f39c3c09f807 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8740) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test e2e-dashboard

kyessenov commented on mixer/adapter/signalfx/tracing.go:
```diff
@@ -160,16 +161,16 @@ func convertInstance(istioSpan *tracespan.Instance) *trace.Span {
 
-	if labels, ok := istioSpan.SpanTags["source.labels"].(map[string]string); ok {
-		for k, v := range labels {
+	if labels, ok := istioSpan.SpanTags["source.labels"].(attribute.StringMap); ok {
```
note: interesting mistake occurring when attribute values are placed into generic maps in instances.

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/e2e-simpleTests-minProfile.sh | 75651d2877a2e1252887d5ca4af0d32f6bcb5e7b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/254) | `/test e2e-simpleTestsMinProfile`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kyessenov: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow//e2e-simpleTests-minProfile.sh | 765c813643d43312413922618f0b577d706e82cd | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-simpleTestsMinProfile/203) | `/test e2e-simpleTestsMinProfile`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 3ecf9c53dd4c60f010475d40e46513cb8b831791 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-e2e-envoyv2-v1alpha3/12021) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | 3ecf9c53dd4c60f010475d40e46513cb8b831791 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/istio-pilot-multicluster-e2e/5747) | `/test istio-pilot-multicluster-e2e`
prow/e2e-dashboard.sh | 3ecf9c53dd4c60f010475d40e46513cb8b831791 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10260/e2e-dashboard/8790) | `/test e2e-dashboard`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test istio-pilot-multicluster-e2e

kyessenov commented:
/test e2e-dashboard

kyessenov commented:
/test istio-pilot-e2e-envoyv2-v1alpha3

kyessenov commented:
The PR is ready.
Friendly request for a review.

kyessenov commented:
Ping @geeknoid @mandarjog 
This PR contains the bit to disable L2 cache, which is one known source of flakiness in mixer tests.

geeknoid commented:
Given all these fixes, do we still need to disable the L2 cache?

geeknoid commented on mixer/adapter/signalfx/tracing.go:
```diff
@@ -28,6 +28,7 @@ import (
 	octrace "go.opencensus.io/trace"
 
 	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/attribute"
```
Adapters aren't supposed to deal with attributes, and as such I think importing this package here indicates a problem.

kyessenov replied:
FYI It is used to import `attribute.StringMap`. 

Here is the problem. Trace span uses a field with a dynamic value in the map values:
```
https://github.com/istio/istio/blob/master/mixer/template/tracespan/tracespan.proto#L109
```

Existing code generation works by places the raw attribute value in the map. That means the string map shows up as a value, and `net.IP` is turned into raw bytes. This is nothing new, the code simply didn't work before since it was casting to a wrong type in both cases.

geeknoid replied:
What is the wire format used for OOP adapters for IP and email addresses?

kyessenov replied:
For OOP, the value holds the protobuf dynamic value: https://github.com/istio/istio/blob/master/mixer/template/tracespan/tracespan_handler_service.proto#L154

This code is for in-process adapter, which doesn't look like it would work out of the box.

kyessenov commented:
@geeknoid I believe we need to add tests to validate L2 cache behavior. Specifically, the original bug could have been caught in a unit-test size set-up.

geeknoid commented on mixer/adapter/signalfx/tracing.go:
```diff
@@ -225,18 +226,20 @@ func convertInstance(istioSpan *tracespan.Instance) *trace.Span {
 
-	if ip, ok := istioSpan.SpanTags["destination.ip"].(net.IP); ok {
-		ips := ip.String()
+	if ip, ok := istioSpan.SpanTags["destination.ip"].([]byte); ok {
```
Why is net.IP being replaced in this code with []byte? We want the adapters to receive the stronger type signatures.

kyessenov replied:
It's a problem with in-process adapter code generation. Dynamic values are turned into raw attribute values.

geeknoid commented on mixer/pkg/il/interpreter/result.go:
```diff
@@ -111,13 +111,10 @@ func (r Result) AsInterface() interface{} {
-				return net.IP(r.vi.([]uint8))
-			}
+		if attribute.CheckType(r.vi) {
+			return r.vi
```
I don't think this is right. Expressions should operate on the logical attribute types, not on the transport types.

kyessenov replied:
Expressions are compiled into IL which erases certain types (e.g. `Interface` type is a big bucket of everything-else). As IL interpreter executes, the inner interface instances are loosely typed from then on. This check is to ensure that the stuff in the result (as produced by foreign functions, for example), is well-typed.

geeknoid commented:
Well, fine, since the in-proc Go adapter model is on the way out I suppose
it doesn't matter.

/lgtm
/approve

On Mon, Dec 10, 2018 at 12:57 PM Kuat <notifications@github.com> wrote:

> *@kyessenov* commented on this pull request.
> ------------------------------
>
> In mixer/adapter/signalfx/tracing.go
> <https://github.com/istio/istio/pull/10260#discussion_r240376975>:
>
> > @@ -28,6 +28,7 @@ import (
>  	octrace "go.opencensus.io/trace"
>
>  	"istio.io/istio/mixer/pkg/adapter"
> +	"istio.io/istio/mixer/pkg/attribute"
>
> For OOP, the value holds the protobuf dynamic value:
> https://github.com/istio/istio/blob/master/mixer/template/tracespan/tracespan_handler_service.proto#L154
>
> This code is for in-process adapter, which doesn't look like it would work
> out of the box.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/10260#discussion_r240376975>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHT3arWngAM7wKSttTI4wmraTeBObks5u3srFgaJpZM4Y_rmV>
> .
>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/10260#issuecomment-445987853" title="Approved">geeknoid</a>*, *<a href="https://github.com/istio/istio/pull/10260#" title="Author self-approved">kyessenov</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[mixer/OWNERS](https://github.com/istio/istio/blob/release-1.1/mixer/OWNERS)~~ [geeknoid,kyessenov]
- ~~[tests/OWNERS](https://github.com/istio/istio/blob/release-1.1/tests/OWNERS)~~ [geeknoid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
